shared/src/event_handling.rs

```diff
@@ -0,0 +1,181 @@
+use anyhow::{Context, Error, Result};
+use ethcontract::contract::{AllEventsBuilder, ParseLog};
+use ethcontract::errors::ExecutionError;
+use ethcontract::{dyns::DynTransport, BlockNumber as Web3BlockNumber, Event as EthcontractEvent};
+use futures::{Stream, StreamExt, TryStreamExt};
+use std::ops::RangeInclusive;
+use web3::Web3;
+
+// We expect that there is never a reorg that changes more than the last n blocks.
+const MAX_REORG_BLOCK_COUNT: u64 = 25;
+// When we insert new trade events into the database we will insert at most this many in one
+// transaction.
+const INSERT_EVENT_BATCH_SIZE: usize = 10_000;
+
+pub struct EventHandler<C: EventRetrieving, S: EventStoring<C::Event>> {
+    contract: C,
+    store: S,
+    last_handled_block: Option<u64>,
+}
+
+/// `EventStoring` is used by `EventHandler` for the purpose of giving the user freedom
+/// in how, where and which events are stored.
+///
+/// # Examples
+/// Databases: might transform, filter and classify which events are inserted
+/// HashSet: For less persistent (in memory) storing, insert events into a set.
+#[async_trait::async_trait]
+pub trait EventStoring<T> {
+    /// Returns ok, on successful execution, otherwise an appropriate error
+    ///
+    /// # Arguments
+    /// * `events` the contract events to be saved by the implementer
+    /// * `range` optionally indicates a particular range of blocks on which to operate.
+    async fn save_events(
+        &self,
+        events: Vec<EthcontractEvent<T>>,
+        range: Option<RangeInclusive<BlockNumber>>,
+    ) -> Result<()>;
+
+    async fn last_event_block(&self) -> Result<u64>;
+}
+
+pub trait EventRetrieving {
+    type Event: ParseLog;
+    fn get_events(&self) -> AllEventsBuilder<DynTransport, Self::Event>;
+    fn web3(&self) -> Web3<DynTransport>;
+}
+
+impl<C, S> EventHandler<C, S>
+where
+    C: EventRetrieving,
+    S: EventStoring<C::Event>,
+{
+    pub fn new(contract: C, store: S, start_sync_at_block: Option<u64>) -> Self {
+        Self {
+            contract,
+            store,
+            last_handled_block: start_sync_at_block,
+        }
+    }
+
+    async fn event_block_range(&self) -> Result<RangeInclusive<BlockNumber>> {
+        let web3 = self.contract.web3();
```

Comment:
I think this was already a cheeky way of getting web3 in the previous implementation, but now we are leaking this hack through the public EventRetrieving interface (by requiring every implementation to implement a method that hands out web3). 

Instead, we could just give the EventHandler component its own web3 instance, this way the EventRetrieving interface is easier to implement in multiple places.

Reply:
Absolutely agree. We had touched on this when refactoring and never came back.

Reply:
> public EventRetrieving interface (by requiring every implementation to implement a method that hands out web3).

Since this is intended to be implemented by contract bindings, I think this is fine, as contract bindings always have a `Web3` instance. Especially given @bh2smith's comment

> specifically implementing EventRetrieving for GPv2Settlement which we intend to remove via a macro

Reply:
To me accessing web3 this way is lying about dependencies. This component needs a web3 to do its logic (in fact it needs a currentBlockFetcher) and if e.g. we wanted to unit test it we could pass in a Fake `AllEventsBuilder` and assert some things about the events that get forwarded.

Reply:
Just to be clear, I also think that passing in a `web3` is the more correct/cleaner approach (or event better `CurrentBlockFetching` if that already exists).

The motivation for not passing it in was just that it was easier and all `EventRetrieving` implementations should be contract bindings and hence have access to a `web3` instance anyway.

Reply:
If its alright with the reviewers, I will leave this as is for now with expectations to address in the following PR where we remove some of boiler plate code referred to here.